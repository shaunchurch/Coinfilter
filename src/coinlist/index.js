// @flow
import CoinListReducer from './coinlist.reducers';
import CoinListContainer from './CoinListCointainer';
import type { Coin } from './coinlist.types';
import CoinListSaga from './coinlist.sagas';

export type { Coin };
export { CoinListReducer, CoinListContainer, CoinListSaga };
