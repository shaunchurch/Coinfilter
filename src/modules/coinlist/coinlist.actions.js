// @flow
import type { Action, ApiCoin } from './coinlist.types';

export const reqCoins = (): Action => {
  return {
    type: 'coinlist/REQ_COINS'
  };
};

export const resCoins = (coins: Array<ApiCoin>): Action => {
  return {
    type: 'coinlist/RES_COINS_SUCCESS',
    payload: coins
  };
};

export const resCoinsFail = (error: Error): Action => {
  return {
    type: 'coinlist/RES_COINS_FAIL',
    error
  };
};

export const doAddHodl = (coinId: string): Action => {
  return {
    type: 'coinlist/ADD_HODL',
    payload: coinId
  };
};

export const doRemoveHodl = (coinId: string): Action => {
  return {
    type: 'coinlist/REMOVE_HODL',
    payload: coinId
  };
};
