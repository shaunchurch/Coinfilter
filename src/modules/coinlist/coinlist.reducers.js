// @flow
import { combineReducers } from 'redux';
import type { ApiCoin, Coin, Action } from './coinlist.types';

type AllState = {
  [id: string]: Coin
};

export const data = (state: AllState = {}, action: Action): AllState => {
  switch (action.type) {
    case 'coinlist/RES_COINS_SUCCESS':
      const coinsObj = {};
      action.payload.map((coin: ApiCoin) => {
        return (coinsObj[coin.id] = typeCoin(coin));
      });
      return coinsObj;
    default:
      return state;
  }
};

export const all = (
  state: Array<string> = [],
  action: Action
): Array<string> => {
  switch (action.type) {
    case 'coinlist/RES_COINS_SUCCESS':
      const coins = action.payload.map((coin: ApiCoin) => {
        const { id } = coin;
        return id;
      });
      return coins;
    default:
      return state;
  }
};

export const hodls = (
  state: Array<string> = [],
  action: Action
): Array<string> => {
  switch (action.type) {
    case 'coinlist/ADD_HODL':
      return state.includes(action.payload)
        ? state
        : [...state, action.payload];
    case 'coinlist/REMOVE_HODL':
      return state.filter(item => item !== action.payload);
    default:
      return state;
  }
};

/**
 * Convert ApiCoin objects with all string values to numbers when
 * the value is a number.
 */
const typeCoin = (coin: ApiCoin): Coin => {
  return {
    '24h_volume_usd': parseInt(coin['24h_volume_usd'], 10),
    available_supply: parseInt(coin.available_supply, 10),
    id: coin.id,
    last_updated: parseInt(coin.last_updated, 10),
    market_cap_usd: parseInt(coin.market_cap_usd, 10),
    name: coin.name,
    percent_change_1h: parseFloat(coin.percent_change_1h),
    percent_change_7d: parseFloat(coin.percent_change_7d),
    percent_change_24h: parseFloat(coin.percent_change_24h),
    price_btc: parseFloat(coin.price_btc),
    price_usd: parseFloat(coin.price_usd),
    rank: parseInt(coin.rank, 10),
    symbol: coin.symbol,
    total_supply: parseInt(coin.total_supply, 10)
  };
};

export default combineReducers({
  data,
  all,
  hodls
});
