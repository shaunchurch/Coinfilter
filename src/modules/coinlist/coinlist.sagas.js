import { all, call, put, takeLatest } from 'redux-saga/effects';
import CoinMarketCap from '../../services/CoinMarketCap';
import type { Action, Coin } from './coinlist.types';

function* watchFetchCoins() {
  yield takeLatest('coinlist/REQ_COINS', fetchCoins);
}

export const fetchCoins: * = function*(action: Action) {
  try {
    const apiCoins: Array<Coin> = yield call(CoinMarketCap.ticker);
    yield put({ type: 'coinlist/RES_COINS_SUCCESS', payload: apiCoins });
  } catch (e) {
    yield put({ type: 'coinlist/RES_COINS_FAILED', message: e.message });
  }
};

export default function* rootSaga() {
  yield all([watchFetchCoins()]);
}
