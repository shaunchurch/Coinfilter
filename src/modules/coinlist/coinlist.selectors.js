// @flow
import type { AppState } from '../../types';
import type { Coin } from './coinlist.types';

/**
 * Select all the coins
 *
 * @param {AppState} state
 */
export const selectCoins = (state: AppState): Array<Coin> => {
  return state.coins.all.map(coin => state.coins.data[coin]);
};

/**
 * Select coins from the user's hodl list
 *
 * @param {AppState} state
 */
export const selectHodls = (state: AppState): Array<Coin> => {
  return state.coins.hodls
    ? state.coins.hodls
        .map(coin => state.coins.data[coin])
        .sort(sortByMarketCap)
    : [];
};

/**
 * Select coins that are not listed in the user's hodl list
 *
 * @param {AppState} state
 */
export const selectOthers = (state: AppState): Array<Coin> => {
  return state.coins.all
    .map(coin => {
      return state.coins.data[coin];
    })
    .filter(coin => {
      return !state.coins.hodls.includes(coin.id);
    })
    .sort(sortByMarketCap);
};

/**
 * Coins by market cap sorting
 *
 * @param {Coin} a
 * @param {Coin} b
 */
const sortByMarketCap = (a: Coin, b: Coin) => {
  if (a.market_cap_usd > b.market_cap_usd) return -1;
  if (a.market_cap_usd < b.market_cap_usd) return 1;
  return 0;
};
