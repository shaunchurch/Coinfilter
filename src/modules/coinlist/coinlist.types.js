// @flow
export type Coin = {
  '24h_volume_usd': number,
  available_supply: number,
  id: string,
  last_updated: number,
  market_cap_usd: number,
  name: string,
  percent_change_1h: number,
  percent_change_7d: number,
  percent_change_24h: number,
  price_btc: number,
  price_usd: number,
  rank: number,
  symbol: string,
  total_supply: number
};

export type ApiCoin = {
  '24h_volume_usd': string,
  available_supply: string,
  id: string,
  last_updated: string,
  market_cap_usd: string,
  name: string,
  percent_change_1h: string,
  percent_change_7d: string,
  percent_change_24h: string,
  price_btc: string,
  price_usd: string,
  rank: string,
  symbol: string,
  total_supply: string
};

export type Action =
  | { type: 'coinlist/REQ_COINS' }
  | { type: 'coinlist/RES_COINS_SUCCESS', payload: Array<ApiCoin> }
  | { type: 'coinlist/RES_COINS_FAIL', error: Error }
  | { type: 'coinlist/ADD_HODL', payload: string }
  | { type: 'coinlist/REMOVE_HODL', payload: string };
