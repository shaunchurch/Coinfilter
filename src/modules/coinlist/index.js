// @flow
import CoinListReducer from './coinlist.reducers';
import CoinListContainer from './CoinListCointainer';
import CoinListSaga from './coinlist.sagas';
import type { Coin } from './coinlist.types';

export type { Coin };
export { CoinListReducer, CoinListContainer, CoinListSaga };
