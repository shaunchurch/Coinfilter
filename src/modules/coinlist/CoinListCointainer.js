// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { reqCoins, doAddHodl, doRemoveHodl } from './coinlist.actions';
import { selectCoins, selectHodls, selectOthers } from './coinlist.selectors';
import CoinList from './components/CoinList';
import type { AppState } from '../../types';
import type { Connector } from 'react-redux';
import type { Coin } from './coinlist.types';

type Props = {
  coins: Array<Coin>,
  others: Array<Coin>,
  hodls: Array<Coin>,
  doAddHodl: Function,
  doRemoveHodl: Function,
  doReqCoins: Function
};

class CoinListContainer extends Component<Props> {
  componentDidMount() {
    this.props.doReqCoins();
  }

  handleAddHodl = (coinId: string) => {
    this.props.doAddHodl(coinId);
  };

  handleRemoveHodl = (coinId: string) => {
    this.props.doRemoveHodl(coinId);
  };

  render() {
    const { hodls, others, coins } = this.props;
    return (
      <div>
        <CoinList
          title="Hodl Coins"
          coins={hodls}
          handleClick={this.handleRemoveHodl}
        />
        <CoinList
          title={'Other Coins'}
          coins={others}
          handleClick={this.handleAddHodl}
        />
      </div>
    );
  }
}

const mapStateToProps = (state: AppState) => {
  return {
    coins: selectCoins(state),
    others: selectOthers(state),
    hodls: selectHodls(state)
  };
};

const mapDispatchToProps = (dispatch: Dispatch) => {
  return {
    doAddHodl: (coinId: string) => {
      dispatch(doAddHodl(coinId));
    },
    doRemoveHodl: (coinId: string) => {
      dispatch(doRemoveHodl(coinId));
    },
    doReqCoins: () => {
      dispatch(reqCoins());
    }
  };
};

const connector: Connector<{}, Props> = connect(
  mapStateToProps,
  mapDispatchToProps
);
export default connector(CoinListContainer);
