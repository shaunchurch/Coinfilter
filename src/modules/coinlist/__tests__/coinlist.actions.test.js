// @flow
import { all } from '../coinlist.reducers';
import { resCoins, reqCoins, doAddHodl } from '../coinlist.actions';
import CoinJSON from '../../../testutils/data/Coin.json';
import ApiCoinJSON from '../../../testutils/data/ApiCoin.json';

describe('coinlist.actions', () => {
  describe('resCoins', () => {
    it('should return a coinlist/RES_COINS_SUCCESS action', () => {
      const actual = resCoins(ApiCoinJSON);
      expect(actual).toMatchSnapshot();
    });
  });

  describe('reqCoins', () => {
    it('should return a coinlist/REQ_COINS action', () => {
      const actual = reqCoins();
      expect(actual).toMatchSnapshot();
    });
  });

  describe('doAddHodl', () => {
    it('should return a coinlist/ADD_HODL action', () => {
      const actual = doAddHodl('bitcoin');
      expect(actual).toMatchSnapshot();
    });
  });
});
