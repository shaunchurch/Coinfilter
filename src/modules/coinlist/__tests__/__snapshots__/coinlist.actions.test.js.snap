// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`coinlist.actions doAddHodl should return a coinlist/ADD_HODL action 1`] = `
Object {
  "payload": "bitcoin",
  "type": "coinlist/ADD_HODL",
}
`;

exports[`coinlist.actions reqCoins should return a coinlist/REQ_COINS action 1`] = `
Object {
  "type": "coinlist/REQ_COINS",
}
`;

exports[`coinlist.actions resCoins should return a coinlist/RES_COINS_SUCCESS action 1`] = `
Object {
  "payload": Array [
    Object {
      "24h_volume_usd": "1511610000.0",
      "available_supply": "16526487.0",
      "id": "bitcoin",
      "last_updated": "1503791971",
      "market_cap_usd": "71929882019.0",
      "name": "Bitcoin",
      "percent_change_1h": "-0.18",
      "percent_change_24h": "-0.45",
      "percent_change_7d": "5.29",
      "price_btc": "1.0",
      "price_usd": "4352.4",
      "rank": "1",
      "symbol": "BTC",
      "total_supply": "16526487.0",
    },
    Object {
      "24h_volume_usd": "514181000.0",
      "available_supply": "94263186.0",
      "id": "ethereum",
      "last_updated": "1503791974",
      "market_cap_usd": "31472781162.0",
      "name": "Ethereum",
      "percent_change_1h": "0.26",
      "percent_change_24h": "0.58",
      "percent_change_7d": "13.11",
      "price_btc": "0.0768412",
      "price_usd": "333.882",
      "rank": "2",
      "symbol": "ETH",
      "total_supply": "94263186.0",
    },
  ],
  "type": "coinlist/RES_COINS_SUCCESS",
}
`;
